In [1]:
import pandas as pd
import numpy as np

In [32]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, matthews_corrcoef
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

In [5]:
data = pd.read_csv('Large_input_with_labels2.csv')

In [10]:
data.columns
data.shape

(2472, 2328)

In [9]:
data['Unnamed: 0']

0          0
1          1
2          2
3          3
4          4
        ... 
2467    2579
2468    2580
2469    2581
2470    2582
2471    2583
Name: Unnamed: 0, Length: 2472, dtype: int64

In [16]:
data = data.drop(columns=['Unnamed: 0'])

In [18]:
data

0  1  2  3  4  5  6  7  8  9  ...      2317       2318       2319  \
0     1  0  0  0  0  0  0  0  0  1  ...  2.041041  14.696329   0.000000   
1     1  1  0  0  0  0  0  0  0  1  ...  1.944782   7.845920   4.909027   
2     1  0  0  0  0  0  0  0  0  1  ...  2.044604  29.700074   4.843949   
3     1  1  1  0  0  0  0  0  0  1  ...  2.008172   8.586421   0.000000   
4     1  1  0  0  0  0  0  0  0  1  ...  1.935553  10.806312   8.290285   
...  .. .. .. .. .. .. .. .. .. ..  ...       ...        ...        ...   
2467  1  1  1  1  0  0  0  0  0  1  ...  2.001864  15.699225  10.519592   
2468  1  1  1  1  0  0  0  0  0  1  ...  2.063046  21.871267  21.871267   
2469  1  1  1  1  0  0  0  0  0  1  ...  2.056623  37.465256  30.626195   
2470  1  1  1  1  0  0  0  0  0  1  ...  2.035607  32.774536  20.025759   
2471  1  1  1  0  0  0  0  0  0  1  ...  2.041022  29.222063   8.679284   

           2320    2321  2322   2323  2324  2325  2326  
0     12.151004   105.0    12 -0.256    52     1     0  
1      2.936893   166.0    11  1.079    50     1     0  
2     21.702655   676.0    28  0.253   102     1     0  
3      6.061732   718.0    25  4.288    92     0     1  
4      0.000000   468.0    21  3.210    76     1     0  
...         ...     ...   ...    ...   ...   ...   ...  
2467   0.000000  3162.0    75  5.294   198     0     1  
2468   0.000000  4759.0    82  3.418   224     0     1  
2469   6.839060  9010.0    91  3.598   260     0     1  
2470  12.748777  9300.0   100  3.551   286     0     1  
2471  15.507074  3655.0    50  3.060   176     1     0  

[2472 rows x 2327 columns]

In [57]:
X = np.array(data.drop(columns = ['2325','2326']) )#and 2326
y = np.array(data['2325'])

In [58]:
y

array([1, 1, 1, ..., 0, 0, 1])

In [59]:
X.shape

(2472, 2325)

In [60]:
kmeans = KMeans(n_clusters=2, random_state=0).fit(X)

In [62]:
sum(kmeans.labels_)

1

In [27]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()

In [64]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, random_state=1)
print("Train proportion: ",len(Xtrain)/len(X))

Train proportion:  0.75


In [65]:
stdscale = StandardScaler()
scaled_train = stdscale.fit_transform(Xtrain[:, 881:])
scaled_test = stdscale.transform(Xtest[:, 881:])

Xtrain = np.concatenate((Xtrain[:, :881], scaled_train), axis=1)
Xtest = np.concatenate((Xtest[:, :881], scaled_test), axis=1)

In [78]:
skmodel = LogisticRegression(class_weight='balanced')
#skmodel = LogisticRegression()
#skmodel = GaussianNB()
skmodel.fit(Xtrain, ytrain)                  
y_model = skmodel.predict(Xtest)   
ytrain_model = skmodel.predict(Xtrain)
print('Train accuracy: ',accuracy_score(ytrain, ytrain_model))
print('Test accuracy: ',accuracy_score(ytest, y_model))
print('F1 score: ',f1_score(ytest, y_model))
print('Matthews correlation coefficient: ',matthews_corrcoef(ytest,y_model))

Train accuracy:  0.8759439050701187
Test accuracy:  0.5679611650485437
F1 score:  0.5686591276252019
Matthews correlation coefficient:  0.13749560921623707


/home/julian/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


Train accuracy:  0.8759439050701187

Test accuracy:  0.5679611650485437

F1 score:  0.5686591276252019

Matthews correlation coefficient:  0.13749560921623707

In [68]:
y.sum()/len(y)

0.48220064724919093

In [69]:
ytrain.sum()/len(ytrain)

0.48220064724919093

In [71]:
ytrain.sum()

894

In [72]:
y.sum()

1192

In [74]:
ytest.sum()/len(ytest)

0.48220064724919093